File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        int ans = 0, n = nums.length, sum = 0;
        Map<Integer, Integer>mp = new HashMap<>();
        mp.put(sum, 1);
        for(int i=0; i<n; i++) {
            sum += nums[i];
            if(mp.containsKey(sum - goal)) {
                ans += mp.get(sum - goal);
            }
            mp.put(sum, mp.getOrDefault(sum, 0) + 1);
        }
        return ans;
    }
}

File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        int ans = 0, n = nums.size();
        unordered_map<int, int>mp;
        mp[0] = 1;
        for(int i=0, sum = 0; i<n; i++) {
            sum += nums[i];
            if(mp.find(sum - goal) != mp.end()) {
                ans += mp[sum - goal];
            }
            mp[sum]++;
        }
        return ans;
    }
};